### ISSUE TITLE
Opus file not playable in VLC

### ISSUE THREAD

**Issue by Antonito** [pion/webrtc#485]
There seems to be an error with the OGG Opus files created.
While they work fine with `ffplay` and within web-browsers, VLC is unable to play them.

`ogginfo`, `opusinfo` report:
```
WARNING: Hole in data (28 bytes) found at approximate offset 4500 bytes. Corrupted Ogg.
...
Several other  "hole in data"
...
ERROR: No Ogg data found in file "output.opus".
Input probably not Ogg.
```

`ffprobe -v error -show_format -show_streams output.opus` report:
```
[STREAM]
index=0
codec_name=opus
codec_long_name=Opus (Opus Interactive Audio Codec)
profile=unknown
codec_type=audio
codec_time_base=1/48000
codec_tag_string=[0][0][0][0]
codec_tag=0x0000
sample_fmt=fltp
sample_rate=48000
channels=2
channel_layout=stereo
bits_per_sample=0
id=N/A
r_frame_rate=0/0
avg_frame_rate=0/0
time_base=1/48000
start_pts=-960
start_time=-0.020000
duration_ts=436800
duration=9.100000
bit_rate=N/A
max_bit_rate=N/A
bits_per_raw_sample=N/A
nb_frames=N/A
nb_read_frames=N/A
nb_read_packets=N/A
DISPOSITION:default=0
DISPOSITION:dub=0
DISPOSITION:original=0
DISPOSITION:comment=0
DISPOSITION:lyrics=0
DISPOSITION:karaoke=0
DISPOSITION:forced=0
DISPOSITION:hearing_impaired=0
DISPOSITION:visual_impaired=0
DISPOSITION:clean_effects=0
DISPOSITION:attached_pic=0
DISPOSITION:timed_thumbnails=0
[/STREAM]
[FORMAT]
filename=output.opus
nb_streams=1
nb_programs=0
format_name=ogg
format_long_name=Ogg
start_time=-0.020000
duration=9.100000
size=49454
bit_rate=43476
probe_score=100
[/FORMAT]
```

Also, a simple C++ program built with `TagLib 1.11.1` reports no error.

**Comment by Demuzx** [comment]
So how is it going?

**Comment by Demuzx** [comment]
Can you make a normal opus writer?

**Comment by Sean-Der** [comment]
Here is a minimal reproduce. Reading the docs on Opus now

```
package main

import (
    "encoding/hex"

    "github.com/pion/rtp"
    "github.com/pion/webrtc/pkg/media/opuswriter"
)

func main() {
    payloads := []string{
        "806f05bc3971200dc84cccec780f5a9013d00b568d076f6408f80328b61bb957ed41c48e0911edc454e84516f5d34b9b6f8063d41a173eab6d82a1fe96db726fda37628b227aee603063fa62759664cfb6a7",
        "806f05bd397123cdc84cccec7886f8ad66e7fbf7bba9bf7770cc5b8501e53ef0e03c17cc9dc7b3d2ffe9c03c8feedfd1270c3f687e8c84386c18fe1bc8b4c8e0b06d47d02311491d9dae5280645c07c06189d3dee8342b84f807e893453766e7c92e006ab97e465b",
        "806f05be3971278dc84cccec781efc8771205a91db5a0b92496b2f1ef57e97e03e762cf748993b3de5277e8c580de70c63b6d80b1a6ffa39b85d801321db65d6b98c6860884bac21ea2a7be630b2c24b05",
        "806f05bf39712b4dc84cccec781aab3d5605b2ff501dd2903cc7c9a77205e0e1fab67903ec2922bd2c61935ea8c7bd78b76b3f3ed5aa338bc0a395718e41c5cb9ebe792b5667a080bac5b49f6017d1e949",
        "806f05c039712f0dc84cccec7818659fdef9a578a444e5483366a6b26317714828c970ccb62f1c8f29eb247a0ef31ccd48078e03620a154904ebba411a4b69d96e53e90c8d12d468f0",
        "806f05c1397132cdc84cccec7815414e61ee4a58a2a2320373b7629a35059731dfdc2b0ce378caca34b0f3f02ae5cec2cefc473c149ec408208c79eba97799aa8959c5cb3cc9ded4d201",
        "806f05c23971368dc84cccec78153a4bc8b511d395cae4d27e1c83459a4c00de23087381dcf5331fd454b5e18622578290bc1965add2a9abae0771c98d4894625d39348f0a271a73dd886659f2fb",
        "806f05c339713a4dc84cccec781582e059ad9f2426b1f57a715b35f5a0b13d30ca4ed16d8eef0b6ee2b461257d3131ad8dac21648a4c856fe24332833205ecd34dd4d4f94920d4512f5db8",
        "806f05c439713e0dc84cccec7815f9b033bc6e1012ab90079de23190881e0841dde3767bce213a5703ce58942a7ea64b8b99db81d1bbca2cc001c15805b87233c317362947976f5b0603ea0fb467dd6f884927c5acdd3875b8ad1bd67d7c5250",
        "806f05c5397141cdc84cccec780410a6525959da90684aca7e2f38cf5c67d4e0e1d3d08bdfef1774ef0391550d444ca09546232e8da7465f82553ae4aa7a8d4e6631a8a5dbe42e20292dc5955216aecfa82bbdb9143efcf1533bf0c6f098",
        "806f05c63971458dc84cccec783cfc184e937832d002abb7c1a8d6bcbe60b6292a7ccbcb307506985a998263724815113c1d6690968115f8f24cee6a720212085e4846bae31ab610e0e2c84166c625c2d58b5535bc624b",
        "806f05c73971494dc84cccec788c4dc67c3c0f11ac68c113e075acfd8a78f99d7e683294240d469d73682967669d4d8d27da98244db718fad55c0677bcbae4851d20e060c9f493693d71c7ff5887236f957e765fde88461da711fb94",
        "806f05c839714d0dc84cccec7889eb3d851ed6bac84d945872edbe480763e7661b7f2606018d592ef05e087d291f0897f99a80d0fae37cc0bdac44e4de53a91b1302d64480f3524d40d4b5cca8a641bccf3909a30cfaa3a9abc63f",
        "806f05c9397150cdc84cccec788909f4f05535a831a326a710920a916859ce2d38a4b78bfd78400cedd60c9b7a909d46e11e87d5ffa34dd7771893faf25ab4ece71ac07845022c7e848f28c9d8e7814ecea01c972a622c7c5cb54c",
        "806f05ca3971548dc84cccec782969af904118c70cfae8e6b8fb3bc93ae3e889f53cb0cca0ef9b1a95fb632ac658a3204931120b8ad0797dd8f82bfd94d24da88f4953f28901d3cc420caa484bf36dc1e9db",
        "806f05cb3971584dc84cccec788729ba58cda3442c89e487bb61de4e9454fb0851cf9886f6fb9384260180ee1a0b1a272b07cd660ec9f3e35151f4dbd8e01e59ca6cd00742c98a75f279bc3be484ca02659e973a890606c453810ba5cef9951892",
        "806f05cc39715c0dc84cccec78878dba989a247f8c9bbf5bafe99dbe92e1b1fb18b3601e5805421a7337c942e80879a4e09aefb2fe97a700cfd45895458380cb08087252c5a8515ad6d60243557ed39b61f76138e911451b3656f599933c",
        "806f05cd39715fcdc84cccec78878e988a1c991f246d7d9f58cff743b609a846711a49b333fd3b5a9bbf3939c984aaac8bcebf48fec403fea6586b328d0d922e1ddad4454457ec1e0f13e28265a88242e5a737a5bcba5b",
        "806f05ce3971638dc84cccec78878493c7f0f7f68f907e721d8f21d2ff4f1c8ac21db4e006d86b392e42487e2f58978c3744532fb8800da2dbceb435fd1ff68ab0e5483711d094a0aef7184428036101e18d",
        "806f05cf3971674dc84cccec7825585a253e8e4aab9a030ecad28aa40fdd308b1ed42f96b234364c7bf5d7ddb675da95a7675ec946ccb186363b626fbf4a403c6765c16801e2c7930feae5e4e9db8c12dfa8f8f868c200ecb88d27",
        "806f05d039716b0dc84cccec78881d5d31ea55214f3c64e4e19df4feb0b0b86f833905d6f52e88973153de1d0bab0cb5bf61072e6de1b7f8a7e411178e56ef1d0640d3037603d9794bbc335a171a83503bdbf0731a6c5180c78fa0b3ebabfed5e6ff76bf66c63b2e519602",
        "806f05d139716ecdc84cccec788982ad5740295ff610c3f2471958ceec3c006bf9abfbad4e4a2f5101e505368e1a8350aced3a89fc60f48f788e3fcc0ccac2f30f5a00dc23becef1f1358ba269ab988c2ccbbdbdae6d100545b1fa67cbeb169a9dc39c33",
        "806f05d23971728dc84cccec7888a6a05c6f33c53e5556e53a40caf4c2e6656e3d0757e4c8eb499944ebec1142189824a3960c56746beaff141cec05c96c0e8e492c2272c387dfee560472adf63c30ea4707d3c15228",
        "806f05d33971764dc84cccec7887f6808a467dda1acf92f1a3ee5c5000c324ccaa9ec5d6804927e0880d042b56d70aff59fa9b911f724cd51c89e9dd0eae24697696bd56dc69c66995f72cc33123855a249b47cfba713e18",
        "806f05d439717a0dc84cccec7887ec9a68856b0f9a27df45e2392725efddecb7573cab7db14333575c43acee2eda3715b75acace4b1e20e683bf5f70d6d405510ec7623b48f968337dafcd197faa8006898ff60df5373a0db878e905",
    }

    opusFile, err := opuswriter.New("output.opus", 48000, 2)
    if err != nil {
        panic(err)
    }

    rtpPacket := &rtp.Packet{}
    for _, p := range payloads {
        data, err := hex.DecodeString(p)
        if err != nil {
            panic(err)
        }

        if err := rtpPacket.Unmarshal(data); err != nil {
            panic(err)
        }

        if err := opusFile.WriteRTP(rtpPacket); err != nil {
            panic(err)
        }
    }

    if err = opusFile.Close(); err != nil {
        panic(err)
    }
}
```

**Comment by Sean-Der** [comment]
I am out of time for tonight, but it looks like checksum calculation for pages is incorrect. When I use hardcoded values for everything (but use our checksum code) it fails.

When I use a hardcoded checksum value, `opusinfo` is able to parse it!

**Comment by Sean-Der** [comment]
This is fixed by 6209597312aef11032b3d56d6f55b9688a4bbf04

@Demuzx mind trying it out when you get a chance? thank you!
